core/src/main/java/org/jdbi/v3/core/mapper/reflect/ConstructorInstanceFactory.java

```diff
@@ -36,4 +51,36 @@ T newInstance(Object... params) {
     public String toString() {
         return constructor.toString();
     }
+
+    private static <T> boolean isRecordConstructor(Constructor<T> factory)
+    {
+        // If there is more than one constructor, ensure that the factory constructor is the canonical constructor.
+        // If so it will need to get the type parameters from the declared fields.
+        if (factory.getParameters().length != Arrays.stream(factory.getDeclaringClass().getDeclaredFields())
+            .filter(field -> !Modifier.isStatic(field.getModifiers()))
+            .count()) {
+            return false;
+        }
+        boolean candidate = false;
+        for (int i = 0; i < factory.getParameters().length; i++) {
+            Parameter parameter = factory.getParameters()[i];
+            Field field = factory.getDeclaringClass().getDeclaredFields()[i];
+            if (!parameter.getName().equals(field.getName()) || !parameter.getType().equals(field.getType())) {
+                return false;
+            }
+            // Check if the generic type information is lost, due to jdk21 record constructor: https://bugs.openjdk.org/browse/JDK-8320575
+            if (parameter.getType().getTypeParameters().length > 0 &&
+                !parameter.getParameterizedType().equals(field.getGenericType())) {
```

Comment:
The second condition could be simplified to `!(parameter.getParametrizedType() instanceof ParametrizedType)`, wdyt?

Reply:
I worry a bit about the behavior if you have different types for your parameter and field, like

```
class Test {
    T field;
    Test(Supplier<T> field) {
        this.field = field.get();
    }
}
```

Should we make the check stronger, like checking `GenericTypes.isSuperType(parameter.getType(), field.getGenericType())`?

Reply:
Update: tried that and the `superType` parameter is erased in the test. The existing comparision catches this. Also for the example above, since Supplier<T> is a different class than T, it wouldn't even get that far since the types differ.

lmk if that's ok.

Reply:
ok, that's fine - what I would like to see is a unit test showing that it works when it should and a negative test showing it doesn't match when it shouldn't